### PR TITLE
Core/Misc: bunch of crashfixes

### PR DESCRIPTION
**Pull request by Viste**
## Changes Proposed:
-  bunch of my crashfixes for Acore, some part of core c-p from TrinityCore

## Issues Addressed:
- maybe Closes some of #7078 and some others

## Tests Performed:
- Build and uses on wowka.su and someother ac servers

## How to Test the Changes:

1. Just everything should work the same, but the server won't crash. In short, nothing should change, only the disconnects will disappear :D

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

File: src/server/game/Spells/Auras/SpellAuras.cpp [hunk 1]
```diff
@@ -846,40 +846,44 @@ void Aura::SetDuration(int32 duration, bool withMods)
...
-}
 
-// xinef: dot's rolling function
-void Aura::RefreshTimersWithMods()
-{
-    Unit* caster = GetCaster();
-    m_maxDuration = CalcMaxDuration();
-    if ((caster && caster->HasAuraTypeWithAffectMask(SPELL_AURA_PERIODIC_HASTE, m_spellInfo)) || m_spellInfo->HasAttribute(SPELL_ATTR5_SPELL_HASTE_AFFECTS_PERIODIC))
-        m_maxDuration = int32(m_maxDuration * caster->GetFloatValue(UNIT_MOD_CAST_SPEED));
+    if(!caster)
```
Comment (Kitzunu) [inline src/server/game/Spells/Auras/SpellAuras.cpp:881]:
```suggestion
    if (!caster)
```

File: src/server/game/Spells/Auras/SpellAuras.cpp [hunk 2]
```diff
@@ -930,6 +934,9 @@ void Aura::SetStackAmount(uint8 stackAmount)
     m_stackAmount = stackAmount;
     Unit* caster = GetCaster();
 
+    if(!caster)
```
[the remark this answers is not in the document]
Comment (Viste) [reply src/server/game/Spells/Auras/SpellAuras.cpp:937]: ШОК! Это же надо блин! ТАКУЮ ОШИБКУ ДОПУСТИЛ! :D

[the remark this answers is not in the document]
Comment (FrancescoBorzi) [reply src/server/game/Spells/Auras/SpellAuras.cpp:937]: the codestyle should be consistent

**Comment by FrancescoBorzi** [comment]
thanks @Viste 

@azerothcore/testers can anyone please test this?

File: src/server/game/Spells/Auras/SpellAuras.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (Viste) [reply src/server/game/Spells/Auras/SpellAuras.cpp:937]: @FrancescoBorzi It's just a joke, don't pay any attention to it.

**Comment by Meltie2013** [comment]
> ## Changes Proposed:
> * bunch of my crashfixes for Acore

@Viste Is it possible to include where you got the code from? Most of the event code is copy and paste from TrinityCore.
Please keep credits to the projects, even if you use the code.

**Comment by Viste** [comment]
> > ## Changes Proposed:
> > 
> > * bunch of my crashfixes for Acore
> 
> @Viste Is it possible to include where you got the code from? Most of the event code is copy and paste from TrinityCore.
> Please keep credits to the projects, even if you use the code.

you mean Most of the AzerothCore code  is copy and paste from TrinityCore.
and i think info in readme and headers, also i don't go to some commit and took code, so i even don't know who is the author of code but i agree its from TC so plz chill man i just re use code from TC in TC fork where TC credits puted in every file so i dont understand why you so angry ;)

**Comment by Meltie2013** [comment]
> you mean Most of the AzerothCore code is copy and paste from TrinityCore.
> and i think info in readme and headers sufficient to satisfy the rights of the right holder, also i don't go to some commit and took code i just go to git and take some needed code... so i even don't know who is the author of code but i agree its from TC so plz chill man i just re use code from TC in TC fork where TC credits puted in every file so i dont understand why you so angry ;)
> if its because i wrote my crashfixes. It's very simple to explain. There was a problem, I did some research. Concluded that the fastest and best way to take code from TC and reuse it in a TC fork.
> I don't see any problem with this. All code ownership rights are already specified in the readme and headers. And mine, there, is precisely the job of investigating and applying the fixes, in the case of using code from TC.
> Okay, bygones, my little right wrestler in a field where only blizzard has rights. I pointed out in the description that part of the code belongs to TC, hope that soothes your wounded ego, just do not cry

Not mad about anything, and you took the statement personally about my asking.
Just saw the PR and was asking if it was possible to mention where you got it from. Not that hard to say where its from.

**Comment by Viste** [comment]
In case you're suddenly interested in my opinion.
When you commit to open source at that point the code ceases to be yours and becomes shared, so all this nonsense about copyrights is just nonsense, hypocrisy and cowardice.

**Comment by Viste** [comment]
> Not mad about anything, and you took the statement personally about my asking.
> Just saw the PR and was asking if it was possible to mention where you got it from. Not that hard to say where its from.

For a reason, I took it personally, for a reason. Perhaps it's the notoriety of TC and their developers. Namely to whine about copyrights. Or maybe I'm a dork and it's translation difficulties :D

**Comment by FrancescoBorzi** [comment]
@Viste @Meltie2013 you both made good points and there is no reason to fight, we all want the same thing here :)

Indeed it's nice to mention the author when possible, but as @Viste said we give credits to TC in any possible way

File: src/common/Utilities/EventProcessor.h [hunk 3]
```diff
@@ -8,58 +8,77 @@
...
 
 #include "Define.h"
+#include "Duration.h"
 
 #include <map>
 
+class EventProcessor;
 // Note. All times are in milliseconds here.
 
 class BasicEvent
```
Comment (Winfidonarleyan) [inline src/common/Utilities/EventProcessor.h:18]:
```suggestion
class EventProcessor;

// Note. All times are in milliseconds here.
class BasicEvent
```

File: src/common/Utilities/EventProcessor.h [hunk 4]
```diff
@@ -8,58 +8,77 @@
...
+        STATE_RUNNING,
+        STATE_ABORT_SCHEDULED,
+        STATE_ABORTED
+        };
+
+    public:
+        BasicEvent()
+            : m_abortState(AbortState::STATE_RUNNING), m_addTime(0), m_execTime(0) { }
+
+        virtual ~BasicEvent() { }                           // override destructor to perform some actions on event removal
```
Comment (Winfidonarleyan) [inline src/common/Utilities/EventProcessor.h:34]:
```suggestion
        virtual ~BasicEvent() { } // override destructor to perform some actions on event removal
```

File: src/common/Utilities/EventProcessor.h [hunk 5]
```diff
@@ -8,58 +8,77 @@
...
+
+        // this method executes when the event is triggered
+        // return false if event does not want to be deleted
+        // e_time is execution time, p_time is update interval
+        virtual bool Execute(uint64 /*e_time*/, uint32 /*p_time*/) { return true; }
```
Comment (Winfidonarleyan) [inline src/common/Utilities/EventProcessor.h:39]:
```suggestion
        [[nodiscard]] virtual bool Execute(uint64 /*e_time*/, uint32 /*p_time*/) { return true; }
```

File: src/server/game/Entities/Creature/TemporarySummon.cpp [hunk 6]
```diff
@@ -16,13 +16,33 @@ TempSummon::TempSummon(SummonPropertiesEntry const* properties, ObjectGuid owner
     Creature(isWorldObject), m_Properties(properties), m_type(TEMPSUMMON_MANUAL_DESPAWN),
     m_timer(0), m_lifetime(0)
 {
-    m_summonerGUID = owner;
+    if (owner)
+        m_summonerGUID = owner;
```
Comment (Winfidonarleyan) [inline src/server/game/Entities/Creature/TemporarySummon.cpp:20]:
```suggestion
    if (owner)
    {
        m_summonerGUID = owner;
    }
            
```

File: src/server/game/Entities/Creature/TemporarySummon.cpp [hunk 7]
```diff
@@ -16,13 +16,33 @@ TempSummon::TempSummon(SummonPropertiesEntry const* properties, ObjectGuid owner
...
+WorldObject* TempSummon::GetSummoner() const
+{
+    return m_summonerGUID ? ObjectAccessor::GetWorldObject(*this, m_summonerGUID) : nullptr;
+}
+
+Unit* TempSummon::GetSummonerUnit() const
+{
+    if (WorldObject* summoner = GetSummoner())
+        return summoner->ToUnit();
+    return nullptr;
```
Comment (Winfidonarleyan) [inline src/server/game/Entities/Creature/TemporarySummon.cpp:33]:
```suggestion
    if (WorldObject* summoner = GetSummoner())
    {
        return summoner->ToUnit();
    }
    
    return nullptr;
```

File: src/server/game/Entities/Creature/TemporarySummon.h [hunk 8]
```diff
@@ -37,7 +37,10 @@ class TempSummon : public Creature
     void RemoveFromWorld() override;
     void SetTempSummonType(TempSummonType type);
     void SaveToDB(uint32 /*mapid*/, uint8 /*spawnMask*/, uint32 /*phaseMask*/) override {}
-    [[nodiscard]] Unit* GetSummoner() const;
+    WorldObject* GetSummoner() const;
+    Unit* GetSummonerUnit() const;
+    Creature* GetSummonerCreatureBase() const;
+    GameObject* GetSummonerGameObject() const;
```
Comment (Winfidonarleyan) [inline src/server/game/Entities/Creature/TemporarySummon.h:43]:
```suggestion
    [[nodiscard]] WorldObject* GetSummoner() const;
    [[nodiscard]] Unit* GetSummonerUnit() const;
    [[nodiscard]] Creature* GetSummonerCreatureBase() const;
    [[nodiscard]] GameObject* GetSummonerGameObject() const;
```

File: src/server/game/Entities/Object/Object.h [hunk 9]
```diff
@@ -457,48 +457,33 @@ struct Position
 
     [[nodiscard]] bool IsPositionValid() const;
 
-    [[nodiscard]] float GetExactDist2dSq(float x, float y) const
+    float GetExactDist2dSq(const float x, const float y) const
```
Comment (Winfidonarleyan) [inline src/server/game/Entities/Object/Object.h:460]:
```suggestion
    [[nodiscard]] float GetExactDist2dSq(const float x, const float y) const
```

File: src/server/game/Spells/Auras/SpellAuras.cpp [hunk 10]
```diff
@@ -846,40 +846,44 @@ void Aura::SetDuration(int32 duration, bool withMods)
...
+    {
+        int32 duration = m_spellInfo->GetMaxDuration();
+        // Calculate duration of periodics affected by haste.
+        if (caster->HasAuraTypeWithAffectMask(SPELL_AURA_PERIODIC_HASTE, m_spellInfo) || m_spellInfo->HasAttribute(SPELL_ATTR5_SPELL_HASTE_AFFECTS_PERIODIC))
+            duration = int32(duration * caster->GetFloatValue(UNIT_MOD_CAST_SPEED));
+        SetMaxDuration(duration);
+        SetDuration(duration);
+    }
+    else
+        SetDuration(GetMaxDuration());
```
Comment (Winfidonarleyan) [inline src/server/game/Spells/Auras/SpellAuras.cpp:866]:
```suggestion
        SetDuration(GetMaxDuration());

```

File: src/server/game/Spells/Auras/SpellAuras.cpp [hunk 11]
```diff
@@ -846,40 +846,44 @@ void Aura::SetDuration(int32 duration, bool withMods)
...
+    if (!caster)
+        return;
 
+    if (withMods && caster)
+    {
+        int32 duration = m_spellInfo->GetMaxDuration();
+        // Calculate duration of periodics affected by haste.
+        if (caster->HasAuraTypeWithAffectMask(SPELL_AURA_PERIODIC_HASTE, m_spellInfo) || m_spellInfo->HasAttribute(SPELL_ATTR5_SPELL_HASTE_AFFECTS_PERIODIC))
+            duration = int32(duration * caster->GetFloatValue(UNIT_MOD_CAST_SPEED));
+        SetMaxDuration(duration);
```
Comment (Winfidonarleyan) [inline src/server/game/Spells/Auras/SpellAuras.cpp:862]:
```suggestion
        SetMaxDuration(duration);

```

**Comment by temperrr** [comment]
```
CONFLICT (content): Merge conflict in src/server/scripts/Spells/spell_dk.cpp
Auto-merging src/server/scripts/Northrend/IcecrownCitadel/boss_the_lich_king.cpp
Auto-merging src/server/scripts/Events/brewfest.cpp
Auto-merging src/server/game/Spells/SpellMgr.cpp
Auto-merging src/server/game/Spells/Spell.h
Auto-merging src/server/game/Spells/Spell.cpp
Auto-merging src/server/game/Spells/Auras/SpellAuras.cpp
Auto-merging src/server/game/Spells/Auras/SpellAuraEffects.cpp
Auto-merging src/server/game/Globals/ObjectMgr.cpp
Auto-merging src/server/game/Entities/Unit/Unit.cpp
Auto-merging src/server/game/Entities/Object/Object.h
CONFLICT (content): Merge conflict in src/server/game/Entities/Object/Object.h
Auto-merging src/server/game/Battlefield/Zones/BattlefieldWG.cpp
Automatic merge failed; fix conflicts and then commit the result.

```

Getting this when doing ` git pull origin pull/7307/head`

**Comment by FrancescoBorzi** [comment]
@Viste please fix the merge conflicts

**Comment by Helias** [comment]
There are some issues with the build

**Comment by Viste** [comment]
@Helias yes but i cant fix it today, only tomorrow

**Comment by Helias** [comment]
no worries at all!

**Comment by Viste** [comment]
fixed sorry only today i found a bit time for it

**Comment by temperrr** [comment]
Tested. 
Builds fine. Didnt notice any crashes without PR.
No errors in worldserver visible.
Tested Midnight (Karazhan boss) (as I saw some changes there) and no crashes there with PR (or without PR)
Ran around monsters that have minions summened, didn't notice anything there either.


As I dont know what caused the crashed linked, I don't know if PR fixed it.
If there is anything in particulair you still want me to test, please say so!

File: src/server/game/Entities/Creature/TemporarySummon.cpp [hunk 12]
```diff
@@ -16,13 +16,39 @@ TempSummon::TempSummon(SummonPropertiesEntry const* properties, ObjectGuid owner
     Creature(isWorldObject), m_Properties(properties), m_type(TEMPSUMMON_MANUAL_DESPAWN),
     m_timer(0), m_lifetime(0)
 {
-    m_summonerGUID = owner;
+    if (owner)
```
Comment (UltraNix) [inline src/server/game/Entities/Creature/TemporarySummon.cpp:19]: Hmm, do we support other objects than unit to be a summoner like TC?

**Review by UltraNix — APPROVED** [review]

File: src/server/game/Entities/Creature/TemporarySummon.cpp (hunk 12 above)
Comment (Viste) [reply src/server/game/Entities/Creature/TemporarySummon.cpp:19]: yes

**Comment by FrancescoBorzi** [comment]
thank you very much @Viste and all people who reviewed/tested this

**Comment by mpfans** [comment]
I have been testing other pr. because the server crashed when I tested this PR last time, I suspended the test. Now I have merged this commit, and I will update my AC core to test it

**Pull request merged**